message-row: Introduce MessageIndicators

melix99 opened the pull request:
Fixes #28.

The PR introduces a new MessageIndicators widget that will contain all the icons and labels that are related to the status of the message (sent/unread icons, timestamp label, edited label, etc), currently only the timestamp and sponsored message text is included. ~This also introduces the MessageLabel which is a widget that contains the label of the related message and also the MessageIndicators widget, which is always placed at the bottom right of the label (the label will always leave a space for it to fit).~

~What's missing is a way of updating the label attributes when the MessageIndicators size changes.~

~EDIT: the MessageIndicators also needs to be included in the MessageSticker widget.~ Done.

~EDIT: see https://github.com/melix99/telegrand/pull/121#issuecomment-1006799962.~

marhkb commented on src/session/content/message_label.rs:
```diff
@@ -0,0 +1,148 @@
+use crate::session::content::MessageIndicators;
+use crate::utils::parse_formatted_text;
+
+const PANGO_SCALE: i32 = 1024;
```
I would suggest putting this in utils.rs as I think we will need this more often.

melix99 replied:
Actually this is a constant defined in pango, but I couldn't figure out where it was in the bindings so I just copy pasted the value from pango. Now I just discovered that there's pango::SCALE in the bindings, so I'll replace with it.

melix99 commented on data/resources/ui/content-message-indicators.ui:
```diff
@@ -0,0 +1,16 @@
+  <template class="ContentMessageIndicators" parent="GtkWidget">
+    <property name="layout-manager">
+      <object class="GtkBoxLayout">
+        <property name="spacing">3</property>
```
Testing this with the message status icons it seems that there is too much spacing, so maybe we can completely remove this.

melix99 replied:
Removing it seems better for the `sent` state:
![image](https://user-images.githubusercontent.com/23530586/138616353-16465ae1-dfc6-4078-aff3-b48faba74143.png)

But not so much for the `read` state:
![image](https://user-images.githubusercontent.com/23530586/138616375-35aaf015-0263-4e5e-970d-96b498e7651b.png)

I guess that's because of the wider `read` icon, not sure what to do here.

melix99 replied:
As discussed on Telegram, no actions are needed there.

melix99 committed:
message-row: Introduce MessageIndicators

It can only display the timestamp or if the message is a sponsored
message for now. This also includes a new MessageLabel widget that
can display the MessageIndicators to the end of the label.

melix99 commented on src/session/content/message_row/media.rs:
```diff
@@ -146,6 +161,24 @@ impl Media {
+        if imp.caption_label.is_visible() {
+            let indicators = self.indicators();
+            if let Some(ref indicators) = indicators {
+                imp.overlay.remove_overlay(indicators);
```
This gives a critical when an existing caption of the media is edited.

```
(telegrand:2): Gtk-CRITICAL **: 01:22:44.060: gtk_overlay_remove_overlay: assertion 'gtk_widget_get_parent (widget) == GTK_WIDGET (overlay)' failed
```

This function needs to be redone in a different way.

melix99 replied:
The changes were a bit more involved, but at least the code is a bit better imho.

melix99 committed:
style: Apply correct media caption margins via css

melix99 committed:
message-media: Create caption label on demand

We don't really need an invisible caption label.

melix99 commented on src/session/content/message_row/label.rs:
```diff
@@ -0,0 +1,241 @@
+        false
+    }
+
+    fn update_label(&self) {
```
We should also update pango attributes after updating the label.

melix99 commented on src/session/content/message_row/label.rs:
```diff
@@ -0,0 +1,241 @@
+impl MessageLabel {
+    pub(crate) fn new(label: &str, indicators: &MessageIndicators) -> Self {
+        glib::Object::new(&[("label", &label), ("indicators", indicators)])
+            .expect("Failed to create MessageIndicators")
```
```suggestion
            .expect("Failed to create MessageLabel")
```

Oops

marhkb commented on src/session/content/message_row/indicators.rs:
```diff
@@ -0,0 +1,135 @@
+    <interface>
+      <template class="MessageIndicators" parent="GtkWidget">
+        <property name="layout-manager">
+          <object class="GtkBoxLayout"/>
```
Could we use a `GtkBinLayout`?

melix99 replied:
Right, I've anticipated things for #109.

marhkb commented on src/session/content/message_row/indicators.rs:
```diff
@@ -0,0 +1,135 @@
+    "#)]
+    pub(crate) struct MessageIndicators {
+        pub(super) model: MessageIndicatorsModel,
+    }
```
I think it's fine to let this be a 1-element tuple struct

marhkb commented on src/session/content/message_row/indicators_model.rs:
```diff
@@ -0,0 +1,114 @@
+    use std::cell::RefCell;
+
+    #[derive(Debug, Default)]
+    pub(crate) struct MessageIndicatorsModel {
+        pub(super) message: RefCell<Option<glib::Object>>,
+    }
```
~This could also be a 1-element tuple struct.~ Maybe not this one because the type `glib::Object` doesn't reveal the purpose.

melix99 replied:
In this specific one I'm sure that there will be more parameters in the near future, most specifically signal id handlers.

marhkb commented on src/session/content/message_row/indicators.rs:
```diff
@@ -0,0 +1,135 @@
+
+    pub(crate) fn set_message(&self, message: glib::Object) {
+        self.imp().model.set_message(message);
+    }
```
Is there a specific reason for this indirection?

melix99 replied:
#316.

marhkb commented on src/session/content/message_row/indicators_model.rs:
```diff
@@ -0,0 +1,114 @@
+use gettextrs::gettext;
```
Am I correct, that this file is related to #316?

melix99 replied:
Yep. It's an easier way to write the "glue code" for the widget. In this particular case it's not that useful, but it will when we start implementing more things like the "edited" text, the "single" reaction in private chats, message status, view count, etc.

marhkb commented on src/session/content/message_row/indicators.rs:
```diff
@@ -0,0 +1,135 @@
+    impl WidgetImpl for MessageIndicators {}
+}
+
+glib::wrapper! {
+    pub(crate) struct MessageIndicators(ObjectSubclass<imp::MessageIndicators>)
+        @extends gtk::Widget;
+}
```
Quoting https://gtk-rs.org/gtk4-rs/git/book/g_object_subclassing.html

> Which is why, as of today, subclassing of GObjects in Rust requires to mention all ancestors and interfaces apart from GObject and GInitiallyUnowned.

Does that mean we should add `@implements GtkAccessible, GtkBuildable, GtkConstraintTarget;`?

melix99 replied:
It's not a requirement. You just don't get the APIs that those interfaces provide. So, it's not really useful until we actually need those APIs.

marhkb commented on src/session/content/message_row/label.rs:
```diff
@@ -0,0 +1,244 @@
+            return true;
+        }
+
+        false
```
```suggestion
        (text_direction == pango::Direction::Rtl && widget_direction == gtk::TextDirection::Ltr)
            || text_direction == pango::Direction::Ltr
                && widget_direction == gtk::TextDirection::Rtl
```

marhkb commented on src/session/content/message_row/label.rs:
```diff
@@ -0,0 +1,244 @@
+                        "Label",
+                        "The label of the widget",
+                        None,
+                        glib::ParamFlags::READWRITE | glib::ParamFlags::EXPLICIT_NOTIFY,
```
```suggestion
                        glib::ParamFlags::READWRITE
                            | glib::ParamFlags::CONSTRUCT
                            | glib::ParamFlags::EXPLICIT_NOTIFY,
```

melix99 replied:
It crashes because by default the label in most places is None (e.g. in the `MessageText`). Anyway, we don't need that flag because we don't really care if the label is not setted in the constructing phase.

marhkb commented on src/session/content/message_row/media.rs:
```diff
@@ -155,24 +175,50 @@ impl Media {
+            imp.content.remove(&caption_label);
+
+            // Drop the label so that it will unparent the indicators
+            drop(caption_label);
```
Why does dropping the value unparent the indicators?

melix99 replied:
Because the reference counter goes to 0, so the widget is disposed and the `dispose` of that widget unparents the indicators.

marhkb replied:
Could you not just call `indicators.unparent();`?

melix99 replied:
Nope. The reason is that the `caption_label` will still be dropped right after that block. So, what it will happen is that you first unparent the indicators from the label, then you set the overlay as it's new parent, but then right after that block the `MessageLabel` will be dropped and then it will call `unparent()` again on the indicators, which it will make the indicators to be out-of-tree.

marhkb replied:
Ok, I see now. This is a bad side effect. We could prevent that generally by iterating over the children in `dispose` and unparent them like we do in some places.
This at least wouldn't violate the Principle of Least Surprise.

melix99 replied:
I don't really like that because `MessageLabel` is actually designed to always have a `MessageIndicators` as its child. So, calling `unparent` to that widget is not really something that makes sense imho, because we always manage the indicators e.g. in the measure phase, which is specifically designed to be managed by children. I find more sensible to just drop the label altogether and just let the `dispose` to properly do all its things and to completely remove all references of the indicators in the message label before reusing it elsewhere.

marhkb replied:
Ok, I get your point. But I would suggest making it a bit more elegant. IMHO, this is too expressive. AFAIK, this line alone
```rust
    // The label will also be dropped so that it will unparent the indicators.
    imp.content.remove(&imp.caption_label.take().unwrap());
```
should drop it implicitly.

Reference: 
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=38bad2ec63d4e19485e89e8a27036222
